<?xml version="1.0"?>
<launch>
    <arg name="camera_name" default="stereo" />
    <!-- video_stream_provider can be a number as a video device or a url of a video stream -->
    <arg name="video_stream_provider" default="/dev/video" />
    <!-- set camera fps to -->
    <arg name="camera_fps" default="15" />
    <!-- set buffer queue size of frame capturing to -->
    <arg name="buffer_queue_size" default="10" />
    <!-- frames per second to query the camera for -->
    <arg name="fps" default="30" />
    <!-- frame_id for the camera -->
    <arg name="frame_id" default="$(arg camera_name)" />
    <!-- By default, calibrations are stored to file://${ROS_HOME}/camera_info/${NAME}.yaml
    To use your own fill this arg with the corresponding url, e.g.:
    "file:///$(find your_camera_package)/config/your_camera.yaml" -->
    <arg name="camera_info_url" default="" />
    <!-- flip the image horizontally (mirror it) -->
    <arg name="flip_horizontal" default="false" />
    <!-- flip the image vertically -->
    <arg name="flip_vertical" default="false" />
    <!-- force width and height, 0 means no forcing -->
    <arg name="width" default="1280"/>
    <arg name="height" default="1024"/>
    <!-- if show a image_view window subscribed to the generated stream -->
    <arg name="visualize" default="false"/>


    <!-- images will be published at /camera_name/image with the image transports plugins (e.g.: compressed) installed -->
    <group ns="$(arg camera_name)">
        <node pkg="video_stream_opencv" type="video_stream_node" name="left" output="screen" required="true">
            <param name="camera_name" type="string" value="image_raw" />
            <param name="video_stream_provider" type="string" value="$(arg video_stream_provider)0" />
            <param name="camera_fps" type="double" value="$(arg camera_fps)" />
            <param name="buffer_queue_size" type="int" value="$(arg buffer_queue_size)" />
            <param name="fps" type="double" value="$(arg fps)" />
            <param name="frame_id" type="string" value="$(arg frame_id)" />
            <param name="camera_info_url" type="string" value="$(arg camera_info_url)" />
            <param name="flip_horizontal" type="bool" value="$(arg flip_horizontal)" />
            <param name="flip_vertical" type="bool" value="$(arg flip_vertical)" />
            <param name="width" type="int" value="$(arg width)" />
            <param name="height" type="int" value="$(arg height)" />
        </node>

        <node if="$(arg visualize)" name="left_image_view" pkg="image_view" type="image_view">
            <remap from="image" to="left/image_raw" />
        </node>
        <node pkg="video_stream_opencv" type="video_stream_node" name="right" output="screen" required="true">
            <param name="camera_name" type="string" value="image_raw" />
            <param name="video_stream_provider" type="string" value="$(arg video_stream_provider)1" />
            <param name="camera_fps" type="double" value="$(arg camera_fps)" />
            <param name="buffer_queue_size" type="int" value="$(arg buffer_queue_size)" />
            <param name="fps" type="double" value="$(arg fps)" />
            <param name="frame_id" type="string" value="$(arg frame_id)" />
            <param name="camera_info_url" type="string" value="$(arg camera_info_url)" />
            <param name="flip_horizontal" type="bool" value="$(arg flip_horizontal)" />
            <param name="flip_vertical" type="bool" value="$(arg flip_vertical)" />
            <param name="width" type="int" value="$(arg width)" />
            <param name="height" type="int" value="$(arg height)" />
        </node>

        <node if="$(arg visualize)" name="right_image_view" pkg="image_view" type="image_view">
            <remap from="image" to="right/image_raw" />
        </node>
    </group>

</launch>
